lsp-mode.el

```diff
@@ -6364,83 +6376,109 @@ returned by COMMAND is available via `executable-find'"
-COMMAND-FN can only be a function that takes a single argument, a
-port number. It should return a command for launches a language server
-process listening for TCP connections on the provided port."
+(defun lsp-resolve-tcp-function (command-fn port)
+  "Get a command from COMMAND-FN.
+COMMAND-FN shall be a function that takes a single integer
+argument, PORT, and yields a command (see
+`lsp-resolve-final-function')."
+  (lsp-resolve-final-function (funcall command-fn port)))
+
+(defun lsp--tcp-command-present (command-fn)
+  "Make a checker-function for COMMAND-FN.
+The result is a function that, if called, checks wether
+COMMAND-FN (as defined in `lsp-tcp-connection') yields a command
+that can be executed."
+  (lambda () (lsp-server-present? (lsp-resolve-tcp-function command-fn 0))))
+
+(defun lsp-tcp-connection (command-fn &optional test-fn)
+  "Return an lsp connection object for TCP servers.
+COMMAND-FN will be called to start the server and shall take a
+single argument, the port, which is an int. It should return a
+command to launch the server (see `lsp-resolve-final-function').
+
+TEST-FN is a function that should yield a truthy value if and
+only if the server can be started. If this is unspecified,
+COMMAND-FN will be called with PORT=0 and its output examined."
   (cl-check-type command-fn function)
   (list
-   :connect (lambda (filter sentinel name environment-fn)
-              (let* ((host "localhost")
-                     (port (lsp--find-available-port host (cl-incf lsp--tcp-port)))
-                     (command (funcall command-fn port))
-                     (final-command (if (consp command) command (list command)))
-                     (_ (unless (executable-find (cl-first final-command))
-                          (user-error (format "Couldn't find executable %s" (cl-first final-command)))))
-                     (process-environment
-                      (lsp--compute-process-environment environment-fn))
-                     (proc (make-process :name name :connection-type 'pipe :coding 'no-conversion
-                                         :command final-command :sentinel sentinel :stderr (format "*%s::stderr*" name) :noquery t))
-                     (tcp-proc (lsp--open-network-stream host port (concat name "::tcp"))))
-
-                ;; TODO: Same :noquery issue (see above)
-                (set-process-query-on-exit-flag proc nil)
-                (set-process-query-on-exit-flag tcp-proc nil)
-                (set-process-filter tcp-proc filter)
-                (cons tcp-proc proc)))
-   :test? (lambda () (executable-find (cl-first (funcall command-fn 0))))))
+   :connect
+   (lambda (filter sentinel name environment-fn)
+     (let* ((host "localhost")
+            (port (lsp--find-available-port host (cl-incf lsp--tcp-port)))
+            (command (funcall command-fn port))
+            (final-command (lsp-resolve-tcp-function command port))
```

Comment:
Is this correct? The `command` is result of `(funcall command-fn port)` already, and you're going to treat it as function again in `lsp-resolve-tcp-function`?